### PR TITLE
[oneMKL] Student's T-test implementation

### PR DESCRIPTION
**Pull request by nsemin**
Description
This sample demonstrates usage of oneMKL Vector Statistics functionality. The question about Student's T-test was raised on Intel forum (https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/Student-s-t-Test/m-p/1245341/highlight/false#M30651) 

Fixes # (issue) - https://jira.devtools.intel.com/browse/MKLD-10816

Type of change

X  New feature (non-breaking change which adds functionality)

How Has This Been Tested?

 X Command Line
To run the test you need to source oneAPI package script and run make -f GNUmakefile (on Linux) or nmake (on Windows). Also checked it with -Wall -Wformat-security -Werror=format-security parameters

### CODE REVIEW

File: Libraries/oneMKL/student_t_test/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+### On a Linux* System
+Run `make` to build and run the sample. Three programs are generated, which illustrate different APIs for random number generation.
+
+You can remove all generated files with `make clean`.
+
+### On a Windows* System
+Run `nmake` to build and run the sample. `nmake clean` removes temporary files.
+
+*Warning*: On Windows, static linking with oneMKL currently takes a very long time, due to a known compiler issue. This will be addressed in an upcoming release.
+## Running the Monte Carlo Pi Estimation Sample
```
Comment (aelizaro) [inline Libraries/oneMKL/student_t_test/README.md:51]: wrong description

Comment (nsemin) [reply Libraries/oneMKL/student_t_test/README.md]: thanks. fixed

**Review by JoeOster — CHANGES_REQUESTED** [review]
There is no full description of what the sample is. There should have been a template issued when the PR was created and we need that filled out. Also, did you submit for approval to include this as a sample?

File: Libraries/oneMKL/student_t_test/README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# `Student's T-test` Sample
+
+Student's T-test shows how to use the oneMKL library's Vector Statistics functionality to to decide if the null hypothesys should be accepted or rejected.
```
Comment (JoeOster) [inline Libraries/oneMKL/student_t_test/README.md:42]: I made some changes to the read me, please verify you are ok with them

**Review by JoeOster — CHANGES_REQUESTED** [review]

**Comment by nsemin** [comment]
> There is no full description of what the sample is. There should have been a template issued when the PR was created and we need that filled out. Also, did you submit for approval to include this as a sample?

Hello Joe,
Can you clarify whose approval is needed?

**Comment by JoeOster** [comment]
> > There is no full description of what the sample is. There should have been a template issued when the PR was created and we need that filled out. Also, did you submit for approval to include this as a sample?
> 
> Hello Joe,
> Can you clarify whose approval is needed?

The complete process for [submitting a new sample](https://github.com/oneapi-src/oneAPI-samples/wiki/Contributing-a-New-Sample) can be found in the wiki

**Comment by nsemin** [comment]
> > > There is no full description of what the sample is. There should have been a template issued when the PR was created and we need that filled out. Also, did you submit for approval to include this as a sample?
> > 
> > 
> > Hello Joe,
> > Can you clarify whose approval is needed?
> 
> The complete process for [submitting a new sample](https://github.com/oneapi-src/oneAPI-samples/wiki/Contributing-a-New-Sample) can be found in the wiki

Thanks|!

**Comment by petercad** [comment]
@JoeOster Looks good to me. This sample is illustrating a new domain in oneMKL that isn't covered by existing examples.

File: Libraries/oneMKL/student_t_test/License.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Copyright Intel Corporation
+
```
[the remark this answers is not in the document]
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/License.txt:2]: Copied from that source

File: Libraries/oneMKL/student_t_test/README.md (hunk 2 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/README.md:42]: Looks good for me

**Review by JoeOster — DISMISSED** [review]

**Comment by JoeOster** [comment]
@pmpeter1 and @tomlenth - please approve this sample if deemed appropriate

**Review by tomlenth — DISMISSED** [review]
The updates to the readme look great.  thanks,

File: Libraries/oneMKL/student_t_test/t_test.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  // Perform computations of mean and variance
+  auto dataset =
+      oneapi::mkl::stats::make_dataset<oneapi::mkl::stats::layout::row_major>(
+          1, n, r);
+  oneapi::mkl::stats::mean(queue, dataset, mean_buf);
+  queue.wait_and_throw();
+  oneapi::mkl::stats::central_moment(queue, mean_buf, dataset, variance_buf);
+  queue.wait_and_throw();
+  // Create Host accessors and check the condition
+  auto mean_acc = mean_buf.template get_access<sycl::access::mode::read>();
```
Comment (pmpeter1) [inline Libraries/oneMKL/student_t_test/t_test.cpp:56]: Can you use the SYCL 2020 style of accessor?

File: Libraries/oneMKL/student_t_test/t_test.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+static const auto expected_std_dev = 1.0f;
+// T-test threshold which corresponds to 5% significance level and infinite
+// degrees of freedom
+static const auto threshold = 1.95996f;
+
+// T-test function with expected mean
+// Returns: -1 if something went wrong, 1 - in case of NULL hypothesis should be
+// accepted, 0 - in case of NULL hypothesis should be rejected
+template <typename RealType>
+std::int32_t t_test(sycl::queue& queue, sycl::buffer<RealType, 1>& r,
```
Comment (pmpeter1) [inline Libraries/oneMKL/student_t_test/t_test.cpp:39]: Does it help to shorten "queue" to "q"  as the variable?

File: Libraries/oneMKL/student_t_test/t_test.cpp [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    res = 0;
+  }
+  return res;
+}
+
+// T-test function with two input arrays
+// Returns: -1 if something went wrong, 1 - in case of NULL hypothesis should be
+// accepted, 0 - in case of NULL hypothesis should be rejected
+template <typename RealType>
+std::int32_t t_test(sycl::queue& queue, sycl::buffer<RealType, 1>& r1,
```
Comment (pmpeter1) [inline Libraries/oneMKL/student_t_test/t_test.cpp:72]: Same question on "queue" -> "q"?

File: Libraries/oneMKL/student_t_test/t_test.cpp [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+          1, n2, r2);
+  oneapi::mkl::stats::mean(queue, dataset1, mean1_buf);
+  queue.wait_and_throw();
+  oneapi::mkl::stats::central_moment(queue, mean1_buf, dataset1, variance1_buf);
+  oneapi::mkl::stats::mean(queue, dataset2, mean2_buf);
+  queue.wait_and_throw();
+  oneapi::mkl::stats::central_moment(queue, mean2_buf, dataset2, variance2_buf);
+  queue.wait_and_throw();
+  // Create Host accessors and check the condition
+  auto mean1_acc = mean1_buf.template get_access<sycl::access::mode::read>();
```
Comment (pmpeter1) [inline Libraries/oneMKL/student_t_test/t_test.cpp:97]: Same question on SYSCL 2020 accessors.

File: Libraries/oneMKL/student_t_test/t_test.cpp [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+  if (argc >= 2) {
+    n_points = std::atol(argv[1]);
+    if (n_points == 0) {
+      n_points = n_samples;
+    }
+  }
+
+  if (argc >= 3) {
+    mean = std::atof(argv[2]);
```
Comment (pmpeter1) [inline Libraries/oneMKL/student_t_test/t_test.cpp:145]: Do you need error checking on "mean" for bad command-line arguments?

File: Libraries/oneMKL/student_t_test/t_test_usm.cpp [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+static const auto expected_std_dev = 1.0f;
+// T-test threshold which corresponds to 5% significance level and infinite
+// degrees of freedom
+static const auto threshold = 1.95996f;
+
+// T-test function with expected mean
+// Returns: -1 if something went wrong, 1 - in case of NULL hypothesis should be
+// accepted, 0 - in case of NULL hypothesis should be rejected
+template <typename RealType>
+std::int32_t t_test(sycl::queue& queue, RealType* r, std::int64_t n,
```
Comment (pmpeter1) [inline Libraries/oneMKL/student_t_test/t_test_usm.cpp:39]: Shorten "queue" -> "q"?

File: Libraries/oneMKL/student_t_test/t_test_usm.cpp [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+  if (argc >= 2) {
+    n_points = std::atol(argv[1]);
+    if (n_points == 0) {
+      n_points = n_samples;
+    }
+  }
+
+  if (argc >= 3) {
+    mean = std::atof(argv[2]);
```
Comment (pmpeter1) [inline Libraries/oneMKL/student_t_test/t_test_usm.cpp:142]: Error check on mean for bad command line?

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

File: Libraries/oneMKL/student_t_test/t_test.cpp (hunk 4 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/t_test.cpp]: changed

File: Libraries/oneMKL/student_t_test/t_test.cpp (hunk 5 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/t_test.cpp]: changed, though it doesn't lessen quantity of lines

File: Libraries/oneMKL/student_t_test/t_test.cpp (hunk 6 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/t_test.cpp]: changed

File: Libraries/oneMKL/student_t_test/t_test.cpp (hunk 7 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/t_test.cpp]: changed

File: Libraries/oneMKL/student_t_test/t_test.cpp (hunk 8 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/t_test.cpp:142]: check was added

File: Libraries/oneMKL/student_t_test/t_test_usm.cpp (hunk 9 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/t_test_usm.cpp:39]: changed

File: Libraries/oneMKL/student_t_test/t_test_usm.cpp (hunk 10 above)
Comment (nsemin) [reply Libraries/oneMKL/student_t_test/t_test_usm.cpp:142]: check was added

**Review by aelizaro — APPROVED** [review]

**Comment by JoeOster** [comment]
@pmpeter1 - have all your concerns been met? If so please approve

**Review by pmpeter1 — APPROVED** [review]

**Pull request merged**